File: zig/life.zig
const std = @import("std");

const GENERATIONS = 1000;
const SHOW_WORK = false;

const stdout = std.io.getStdOut().writer();
// don't know where to get this from
const MAX_I32 = @as(i32, 0x7fffffff);
const MIN_I32 = @as(i32, -0x80000000);

const allocator = std.heap.c_allocator;

const Coord = struct {
    x: i32,
    y: i32,
};
const LiveSet = std.AutoHashMap(Coord, void);

const Destiny = enum {
    Live,
    Die,
};
const Change = struct {
    coord: Coord,
    destiny: Destiny,
};
const ChangeSet = std.AutoHashMap(Change, void);

// shorthand to make a Coord
fn c(x: i32, y: i32) Coord {
    return Coord{ .x = x, .y = y };
}
const r_pentomino = [_]Coord{ c(0, 0), c(0, 1), c(1, 1), c(-1, 0), c(0, -1) };

fn applyChanges(alive: LiveSet, changes: ChangeSet) !LiveSet {
    var result = LiveSet.init(allocator);
    try result.ensureCapacity(alive.count() + changes.count());
    var iter = alive.keyIterator();
    while (iter.next()) |next| {
        try result.put(next.*, {});
    }
    var change = changes.keyIterator();
    while (change.next()) |next| {
        if (next.destiny == Destiny.Live) {
            try result.put(next.coord, {});
        }
        if (next.destiny == Destiny.Die) {
            _ = result.remove(next.coord);
        }
    }
    return result;
}

const BoundingBox = struct {
    lowerLeft: Coord,
    upperRight: Coord,
};
fn boundingBox(alive: LiveSet) BoundingBox {
    var minx = MAX_I32;
    var miny = MAX_I32;
    var maxx = MIN_I32;
    var maxy = MIN_I32;
    var iter = alive.keyIterator();
    while (iter.next()) |next| {
        minx = std.math.min(minx, next.x);
        maxx = std.math.max(maxx, next.x);
        miny = std.math.min(miny, next.y);
        maxy = std.math.max(maxy, next.y);
    }
    return BoundingBox{ .lowerLeft = c(minx, miny), .upperRight = c(maxx, maxy) };
}

fn printGeneration(alive: LiveSet) !void {
    try stdout.print("\x1b[2J\x1b[;H", .{});
    const bbox = boundingBox(alive);
    var y = bbox.upperRight.y;
    while (y >= bbox.lowerLeft.y) {
        var x = bbox.lowerLeft.x;
        while (x <= bbox.upperRight.x) {
            if (alive.contains(c(x, y))) {
                try stdout.print("@", .{});
            } else {
                try stdout.print(" ", .{});
            }
            x = x + 1;
        }
        try stdout.print("\n", .{});
        y = y - 1;
    }
    const humanWaitMillis: i32 = 1000 / 30;
    std.time.sleep(humanWaitMillis * 1000000);
}

const neighborOffsets = [_][2]i32{
    [_]i32{ -1, -1 },
    [_]i32{ -1, 0 },
    [_]i32{ -1, 1 },

    [_]i32{ 0, -1 },
    [_]i32{ 0, 1 },

    [_]i32{ 1, -1 },
    [_]i32{ 1, 0 },
    [_]i32{ 1, 1 },
};

fn computeNeighbors(changes: ChangeSet) !LiveSet {
    var result = LiveSet.init(allocator);
    try result.ensureCapacity(changes.count() * 8);
    var iter = changes.keyIterator();
    while (iter.next()) |next| {
        for (neighborOffsets) |pair| {
            const x = pair[0];
            const y = pair[1];
            try result.put(c(next.coord.x + x, next.coord.y + y), {});
        }
    }
    return result;
}

fn neighborCount(alive: LiveSet, next: Coord) i32 {
    var result = @as(i32, 0);
    for (neighborOffsets) |pair| {
        const x = pair[0];
        const y = pair[1];
        if (alive.contains(c(next.x + x, next.y + y))) {
            result = result + 1;
        }
    }
    return result;
}

fn computeChanges(alive: LiveSet, changes: ChangeSet) !ChangeSet {
    var result = ChangeSet.init(allocator);
    var neighbors = try computeNeighbors(changes);
    try result.ensureCapacity(neighbors.count() / 2);
    defer neighbors.deinit();
    var iter = neighbors.keyIterator();
    while (iter.next()) |next| {
        const count = neighborCount(alive, next.*);
        if (count == 3) {
            if (!alive.contains(next.*)) {
                try result.put(Change{ .coord = next.*, .destiny = Destiny.Live }, {});
            }
        } else if (count != 2) {
            if (alive.contains(next.*)) {
                try result.put(Change{ .coord = next.*, .destiny = Destiny.Die }, {});
            }
        }
    }
    return result;
}

fn run() !void {
    var alive = LiveSet.init(allocator);
    defer alive.deinit();
    var changes = ChangeSet.init(allocator);
    defer changes.deinit();
    for (r_pentomino) |*coord, i| {
        const change = Change{ .coord = coord.*, .destiny = Destiny.Live };
        try changes.put(change, {});
    }
    var start = std.time.milliTimestamp();
    var i = @as(i32, 0);
    while (i < GENERATIONS) {
        var updated = applyChanges(alive, changes);
        alive.deinit();
        alive = try updated;
        if (SHOW_WORK) {
            try printGeneration(alive);
        }
        var nextGen = computeChanges(alive, changes);
        changes.deinit();
        changes = try nextGen;
        i = i + 1;
    }
    const diff = std.time.milliTimestamp() - start;
    try stdout.print("{d} generations per second\n", .{GENERATIONS / (@intToFloat(f32, diff) / 1000.0)});
}

pub fn main() !void {
    if (SHOW_WORK) {
        try run();
    } else {
        var i: i32 = 0;
        while (i < 5) {
            try run();
            i = i + 1;
        }
    }
}

File: swift/life.swift
import Foundation

let SHOW_WORK = false;
let GENERATIONS = 1000;

struct Coord : Hashable {
    var x = 0
    var y = 0

    static func == (lhs: Coord, rhs: Coord) -> Bool {
        return lhs.x == rhs.x && lhs.y == rhs.y;
    }

    func hash(into hasher: inout Hasher) {
        hasher.combine(x)
        hasher.combine(y)
    }
    
}
enum Destiny { case live, die }
struct Change : Hashable {
    var coord: Coord;
    var destiny: Destiny;
    
    static func == (lhs: Change, rhs: Change) -> Bool {
        return lhs.coord == rhs.coord && lhs.destiny == rhs.destiny;
    }

    func hash(into hasher: inout Hasher) {
        hasher.combine(coord)
        hasher.combine(destiny)
    }
}
struct BoundingBox {
    let lowerLeft : Coord;
    let upperRight : Coord;
}
let TINYBOX = BoundingBox(lowerLeft: Coord(x: Int.max, y: Int.max),
                          upperRight: Coord(x: Int.min, y: Int.min));


func applyUpdates(_ liveSet: Set<Coord>, _ updates: Set<Change>) -> Set<Coord> {
    let toLive = updates.filter { $0.destiny == .live } .map { $0.coord };
    let toDie = updates.filter { $0.destiny == .die } .map { $0.coord };
    return liveSet.subtracting(toDie).union(toLive);
}

func expandBox(_ box: BoundingBox, _ coord: Coord) -> BoundingBox {
    let lowerLeft = box.lowerLeft;
    let upperRight = box.upperRight;
    return BoundingBox(lowerLeft: Coord(x: min(coord.x, lowerLeft.x),
                                        y: min(coord.y, lowerLeft.y)),
                       upperRight: Coord(x: max(coord.x, upperRight.x),
                                         y: max(coord.y, upperRight.y)));
}

func printBoard(_ liveSet: Set<Coord>) {
    print("\u{1b}[2J\u{1b}[;H", terminator: "")
    let bbox = liveSet.reduce(TINYBOX, expandBox);
    for y in stride(from: bbox.upperRight.y, to: bbox.lowerLeft.y - 1, by: -1) {
        for x in bbox.lowerLeft.x ... bbox.upperRight.x {
            let coord = Coord(x: x, y: y)
            if liveSet.contains(coord) {
                print("@", terminator: "")
            } else {
                print(" ", terminator: "")
            }
        }
        print()
    }
    Thread.sleep(forTimeInterval: 1/30.0)
}

let neighborOffsets = [
  (-1, -1), (-1, 0), (-1, 1),
  ( 0, -1),          ( 0, 1),
  ( 1, -1), ( 1, 0), ( 1, 1)
]

func eight(_ coord: Coord) -> Set<Coord> {
    return Set(neighborOffsets.map { Coord(x: coord.x + $0.0,
                                           y: coord.y + $0.1) })
}

func computeNeighbors(_ changes: Set<Change>) -> Set<Coord> {
    return Set(changes.flatMap { eight($0.coord) })
}

func countNeighbors(_ liveSet: Set<Coord>, _ coord: Coord) -> Int {
    return eight(coord).filter { liveSet.contains($0) } .count;
}

func computeChanges(_ liveSet: Set<Coord>, _ neighbors: Set<Coord>) -> Set<Change> {
    var result : Set<Change> = [];
    for coord in neighbors {
        let n = countNeighbors(liveSet, coord)
        if (n != 2) {
            if (n == 3) {
                if (!liveSet.contains(coord)) {
                    result.insert(Change(coord: coord, destiny: .live))
                }
            } else {
                if (liveSet.contains(coord)) {
                    result.insert(Change(coord: coord, destiny: .die))
                }
            }
        }
    }
    return result;
}

func now() -> TimeInterval {
    return NSDate().timeIntervalSince1970;
}

func run() {
    let start = now()
    let rPentomino = [(0, 0), (0, 1), (1, 1), (-1, 0), (0, -1)]
    var liveSet: Set<Coord> = [];
    var updates = Set(rPentomino.map{
        Change(coord: Coord(x: $0.0, y: $0.1), destiny: .live)
                      })
    for _ in 1...GENERATIONS {
        liveSet = applyUpdates(liveSet, updates)
        if (SHOW_WORK) {
            printBoard(liveSet);
        }
        let neighbors = computeNeighbors(updates)
        updates = computeChanges(liveSet, neighbors)
    }
    let diff = now() - start;
    print(String(format: "%.2f generations per second", Double(GENERATIONS) / diff))
}

func main() {
    if (SHOW_WORK) {
        run()
    } else {
        for _ in 1...5 {
            run()
        }
    }
}

main()
